deep_qa/data/instances/text_classification/frame_instance.py

```diff
@@ -0,0 +1,174 @@
+#
+#    raw input line:
+#            "event:plant absorb water###participant:water###agent:plant###finalloc:soil" TAB "agent:plant"
+#
+#    formatted to:
+#           ["plant", "unk", "unk", "unk", "unk", "plant absorb water",
+#            "soil", "unk", "unk", "unk", "unk", "unk",
+#            "unk", "unk", "unk", "unk", "unk", "unk",
+#            "unk", "unk", "unk", "unk", "unk", "unk",
+#            "unk", "unk", "water"]
+#
+#    padded to a fixed length of say, 5 words per phrase.
+#           ["plant P P P P", "unk P P P P", "unk P P P P", "unk P P P P", "unk P P P P", "plant absorb water P P ",
+#             ...
+
+# TODO PR request for having these in the json as an application specific content
+# the slotnames can vary according to different end applications, e.g., a HowTo tuple, OpenIE tuple ...
+SLOTNAMES_ORDERED =["agent", "beneficiary", "causer", "context", "definition", "event",
+     "finalloc", "headverb", "initloc", "input", "output", "manner",
+     "patient", "resultant", "timebegin", "timeend", "temporal", "hierarchical",
+     "similar", "contemporary", "enables", "mechanism", "condition", "purpose",
+     "cause", "openrel", "participant"]
+UNKNOWN_SLOTVAL = "unk"  # making an open world assumption, we do not observe all the values
+QUES_SLOTVAL = "ques"  # this slot in the frame must be queried/completed.
+MAX_PHRASE_LEN = 6
+
+class FrameInstance(TextInstance):
+
+    """
+    A FrameInstance is a kind of TextInstance that has text in multiple slots. This generalizes a FrameInstance.
+    """
+    def __init__(self,
+                 # the input can have only few slots and not all and in any random order.
+                 # e.g., "event:plant absorb water###participant:water###agent:plant###finalloc:soil" TAB "agent:plant"
+                 text: List[str],
+                 # output is a phrase
+                 label: str=None):
+        super(FrameInstance, self).__init__(label)
+        self.text = text  # "event:plant absorb water###participant:water###agent:plant" TAB "agent:plant"
+
+    def __str__(self):
+        return 'FrameInstance( [' + ',\n'.join(self.text) + '] , ' + str(self.label) + ')'
+
+    @overrides
+    # accumulate words from each slot's phrase
+    def words(self) -> Dict[str, List[str]]:
+        words = []
+        for phrase in self.text:  # phrases
+            phrase_words = self._words_from_text(phrase)
+            words.extend(phrase_words['words'])
+        label_words = self._words_from_text(self.label)
+        words.extend(label_words['words'])
+        return {"words": words}
+
+    @staticmethod
+    # s: "participant:water" => name=participant, val=water
+    def slot_from(s: str, kv_separator: str=":"):
+        kv = s.split(kv_separator)
+        return {'name': kv[0], 'val': kv[1]}
+
+    @staticmethod
+    # s: "event:plant absorb water###participant:water" TAB "participant:water"
+    def unpack_input(s: str, kv_separator: str="\t"):
```

Comment:
Data processing in between the logic renders it unreadable. Small functions make the functionality easier to read, and debug.